### PR TITLE
issue #6142 Fixed prometheusexecreceiver command line processing on Windows

### PR DESCRIPTION
**Pull request by bertysentry**
**Description:**
Fixed execution of commands in `prometheusexecreceiver` on Windows

**Link to tracking Issue:** Issue #6142 

**Testing:** N/A

**Documentation:** N/A

### CODE REVIEW

File: receiver/prometheusexecreceiver/subprocessmanager/manager.go
Comment (MovieStoreGuy) [inline receiver/prometheusexecreceiver/subprocessmanager/manager.go:34]: Can I suggest the following:

```golang
var childProcess *exec.Cmd
switch runtime.GOOS {
case "windows":
   childProcess = exec.Command(proc.Command)
default:
    args, err = shellqoute.Split(proc.Command)
    if err != nil {
         return 0, err
    }
    childProcess = exec.Command(args[0], args[1:]...)
}
```

If in future there needs to be a seperate way for darwin based systems, it would be easy to add in.

Comment (bertysentry) [reply receiver/prometheusexecreceiver/subprocessmanager/manager.go]: Good idea, will do that!

**Comment by bertysentry** [comment]
@MovieStoreGuy @owais All fixed, please review 😊

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by bertysentry** [comment]
Build is apparently successful. How to get the *lint*, *unittest* and *integration-tests* reports? @MovieStoreGuy @owais  Thank you!

**Comment by MovieStoreGuy** [comment]
Since it is your first time committing to the project, it requires a maintainer to run the build for you.

You can run the tests locally using `make test` (don't hold me to that, still yet to consume my first coffee of the day) and that will do what the CI is doing locally. 

You may need to run the add tools target as well but I reference that I am not caffeinated enough to brain function well.

**Comment by bertysentry** [comment]
Thank you @MovieStoreGuy 

The load test for `TestIdleMode` fails with the below error message:

```
RAM consumption is 73 MiB, max expected is 70 MiB
```

The load test for `TestBallastMemory|TestLog10kDPS` fails with no error message (in fact, all tests are reported as OK).

Anyway, the changes brought by this PR have zero effect on the performance of the Collector whatsoever. The performance tests are a bit flaky, I suspect. Maybe we should add a bit more margin there, but this is out of scope for this PR.

Anyway, if you guys are okay with the changes, please merge 😊

**Comment by MovieStoreGuy** [comment]
@bertysentry , typically in these times I suggest rebasing against main

**Comment by bertysentry** [comment]
@MovieStoreGuy Rebased and squashed, so it's all clean. We'll see how the build goes... Fingers crossed!

**Comment by bertysentry** [comment]
@MovieStoreGuy All checks are now successful. If you are okay with the proposed change, please merge (I need this in my project to save the world 😎). Thank you in advance!

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
@tigrannajaryan, @bogdandrutu, @jpkrohling 

can I ask you to merge this in?

**Comment by tigrannajaryan** [comment]
@keitwb can you take a look as a codeowner?

**Comment by tigrannajaryan** [comment]
@bertysentry I don't have a Windows machine to verify this on. Will you be able to add an automated test that verifies it on Windows and Linux?

**Comment by bertysentry** [comment]
OMG, this changes doesn't fix anything at all, it actually breaks all process execution on Windows! The origin of the problem comes from https://github.com/golang/go/issues/17149 which has been open for 5 years, and still not fixed! Kind of a shame...

I'll see how if we can work around the problem (at least not parse the command line as if it was a Linux bash command when running on Windows)

**Comment by bertysentry** [comment]
Sorry for the initial broken changes, that was my first go at Go (ha!). Added unit tests, we should be good now. @MovieStoreGuy You can re-review now (sorry for the wasted time!)

Also, Windows-specific implementation of the *SysProcAttr* structure forced me in providing 2 separate implementations of an *ExecCommand()* function, in a 2 separate files: one compiled on Windows only, the other on other systems.

**Comment by bertysentry** [comment]
@tigrannajaryan @MovieStoreGuy Rebased on main and squash unnecessary broken commit. Tests as passing on both Linux and Windows. Aaaand just discovered the use of `make fmt` and `make lint`... 🙄

File: receiver/prometheusexecreceiver/testdata/end_to_end_metrics_test/test_prometheus_exporter windows.go
Comment (tigrannajaryan) [inline receiver/prometheusexecreceiver/testdata/end_to_end_metrics_test/test_prometheus_exporter windows.go:1]: This file appears to be a copy of test_prometheus_exporter.go. If that's the case instead of duplicating it export `server` func and use here.

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, appears to do the right command line passing on Windows and doesn't change anything on non-Windows platforms.
Need a codeowner review to merge.

File: receiver/prometheusexecreceiver/testdata/end_to_end_metrics_test/test_prometheus_exporter windows.go
Comment (bertysentry) [reply receiver/prometheusexecreceiver/testdata/end_to_end_metrics_test/test_prometheus_exporter windows.go]: And actually, the space in the file name breaks `make checklicense` so I entirely removed this file and designed new unit tests to cover the issue that is being fixed here. Sorry for the repeated issues, I'm still struggling with the pipeline here... (issues that I discovery only once run as GitHub Actions and that should be caught earlier...)

**Comment by bertysentry** [comment]
`go test`, `make lint` and `make checklicense` are now all passing. I expect things to finally go smoothly in the GitHub build. Let me now, thank you @MovieStoreGuy @tigrannajaryan !

File: receiver/prometheusexecreceiver/subprocessmanager/manager_test.go
Comment (tigrannajaryan) [inline receiver/prometheusexecreceiver/subprocessmanager/manager_test.go:128]: Nit: do we care about this case? This seems to verify that `test_crasher.go` correctly implements invalid command line. I think this is superfluous, we are essentially testing the test code :-)

**Review by tigrannajaryan — COMMENTED** [review]
LGTM, can merge after @keitwb review as codeowner.

File: receiver/prometheusexecreceiver/subprocessmanager/manager_test.go
Comment (bertysentry) [reply receiver/prometheusexecreceiver/subprocessmanager/manager_test.go:128]: Yes this case is important to determine whether the `"1 2 3"` argument is processed as 1 single argument or not, and therefore that `0` is processed as the 2nd argument (which determines the exit code of **test_crasher.go**.

This test will fail on Windows without the change introduced in this PR.

**Comment by bertysentry** [comment]
Load tests are failing with this error message (again):

```
RAM consumption is 72 MiB, max expected is 70 MiB
```

Should I rebase against **main** again? Will that magically fix the build issue?

File: receiver/prometheusexecreceiver/subprocessmanager/manager_test.go
Comment (tigrannajaryan) [reply receiver/prometheusexecreceiver/subprocessmanager/manager_test.go:128]: Ah, I see, I misunderstood what the test does. Perhaps add a comment to explain, it was not obvious to me.

**Comment by tigrannajaryan** [comment]
> Load tests are failing with this error message (again):
> 
> ```
> RAM consumption is 72 MiB, max expected is 70 MiB
> ```
> 
> Should I rebase against **main** again? Will that magically fix the build issue?

Yeah, the limit is too low, we are hitting it. I will re-run and see if it passes. If not will need to wait for this to be merged: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6377

File: receiver/prometheusexecreceiver/subprocessmanager/manager_test.go
Comment (bertysentry) [reply receiver/prometheusexecreceiver/subprocessmanager/manager_test.go]: Added comments in a new commit. All tests seem to pass ok now, thank you!

**Comment by bertysentry** [comment]
@keitwb As code owner, please review. It's been reviewed already and all tests now pass. Thank you in advance!

**Comment by bertysentry** [comment]
@tigrannajaryan @MovieStoreGuy Do you guys have any contact of a maintainer who could merge this? Thank you!

**Comment by jpkrohling** [comment]
@bertysentry, looks like @keitwb hasn't reviewed this one yet. Once that is done, I can give a final review and merge it.

**Review by keitwb — APPROVED** [review]
I don't really know if that is the best workaround or not but given the annoying bug with Go I guess it will suffice if you have confirmed it with multiple types of params.

The main thing that concerns me with the code itself is what if the `commandLine` has a double quote in it?  Then it will be a nested double quote in the `/C \"` thing.  Please confirm that is ok.

**Comment by bertysentry** [comment]
> The main thing that concerns me with the code itself is what if the `commandLine` has a double quote in it? Then it will be a nested double quote in the `/C \"` thing. Please confirm that is ok.

Yes I do confirm this is okay and tested (it's also in the unit tests, where the last test executes the `"go" "run" "testdata/test_crasher.go" "4" "0"` (with quotes everywhere, and notably at the very beginning), and it works as expected.

From `CMD.EXE /?`:

> If /C or /K is specified, then the remainder of the command line after
> the switch is processed as a command line, where the following logic is
> used to process quote (") characters:
> 
>     1.  If all of the following conditions are met, then quote characters
>         on the command line are preserved:
> 
>         - no /S switch
>         - exactly two quote characters
>         - no special characters between the two quote characters,
>           where special is one of: &<>()@^|
>         - there are one or more whitespace characters between the
>           two quote characters
>         - the string between the two quote characters is the name
>           of an executable file.
> 
>     2.  Otherwise, old behavior is to see if the first character is
>         a quote character and if so, strip the leading character and
>         remove the last quote character on the command line, preserving
>         any text after the last quote character.

When we have a quote in the command line, then we have more than 2 quotes in the CMD command, so we fall in the "old" behavior described in point 2 above.

Thank you for the review!

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Thanks for the fix @bertysentry !

**Pull request merged**
